cdap-sentry/cdap-sentry-extension/cdap-sentry-binding/src/main/java/co/cask/cdap/security/authorization/sentry/binding/SentryAuthorizer.java

```diff
@@ -200,17 +206,67 @@ public void removeRoleFromPrincipal(Role role, Principal principal) throws Excep
 
   @Override
   public void enforce(EntityId entityId, Principal principal, Set<Action> actions) throws Exception {
-    Preconditions.checkArgument(Principal.PrincipalType.USER == principal.getType(), "The given principal '%s' is of " +
-      "type '%s'. In Sentry authorization checks can only be performed on principal type " +
-      "'%s'.", principal.getName(), principal.getType(), Principal.PrincipalType.USER);
-    Set<Action> allowedActions = binding.getActions(principal, entityId);
-    if (allowedActions.isEmpty()) {
+    Preconditions.checkArgument(
+      Principal.PrincipalType.USER == principal.getType(),
+      "Only support principal type %s for authorization, given principal %s is of type %s",
+      Principal.PrincipalType.USER, principal.getName(), principal.getType());
+
+    Set<WildcardPolicy> policies = binding.getPolicies(principal);
+    LOG.debug("Got policies {} for principal {}, entity {} and actions {}", policies, principal, entityId, actions);
+    if (policies.isEmpty()) {
+      throw new UnauthorizedException(principal, actions, entityId);
+    }
+
+    List<Authorizable> authorizables = new ArrayList<>();
+    binding.toAuthorizables(entityId, authorizables);
+
+    Set<ActionFactory.Action> checkActions = binding.toSentryActions(actions);
+
+    // Check each action against each policy
```

Comment:
From what I understand, `AuthBinding` is just an adapter to convert CDAP objects into Sentry objects and talk to Sentry. The business logic on how to enforce authorization should happen in `SentryAuthorizer`